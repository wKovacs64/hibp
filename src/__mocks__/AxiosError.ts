/** @internal */
export default class AxiosError extends Error {
  constructor(response) {
    super();
    this.response = response;
  }
}
