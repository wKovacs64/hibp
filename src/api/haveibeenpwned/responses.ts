/**
 * Known potential responses from the remote API.
 *
 * Unfortunately, the API does not send a decent human-readable message back
 * with each response, but they are documented on the website:
 * https://haveibeenpwned.com/api/v3#ResponseCodes
 *
 * These objects simply provide a mapping between the HTTP response status code
 * and the corresponding human-readable message so we can throw a more
 * descriptive error for the consumer. (They are also leveraged in our tests.)
 */

import type { ResponseBody } from './types';

/** @internal */
export interface HaveIBeenPwnedApiResponse {
  headers: Map<string, string>;
  status: number;
  statusText?: string;
  body?: ResponseBody;
}

const emptyHeaders = new Map<string, string>();

/** @internal */
export const OK: HaveIBeenPwnedApiResponse = {
  headers: emptyHeaders,
  status: 200,
};

/** @internal */
export const BAD_REQUEST: HaveIBeenPwnedApiResponse = {
  headers: emptyHeaders,
  status: 400,
  statusText:
    'Bad request â€” the account does not comply with an acceptable format.',
};

/**
 * This response has unique behavior. For some reason, the API includes an
 * object in the response body for this one, containing a human-readable
 * message. Manually populating the message here purely for use in tests.
 *
 * @internal
 */
export const UNAUTHORIZED: HaveIBeenPwnedApiResponse = {
  headers: emptyHeaders,
  status: 401,
  body: {
    statusCode: 401,
    message: 'Access denied due to missing hibp-api-key.',
  },
};

/** @internal */
export const FORBIDDEN: HaveIBeenPwnedApiResponse = {
  headers: emptyHeaders,
  status: 403,
  statusText: 'Forbidden - access denied.',
};

/** @internal */
export const BLOCKED: HaveIBeenPwnedApiResponse = {
  headers: new Map([['cf-ray', 'someRayId']]),
  status: 403,
};

/** @internal */
export const NOT_FOUND: HaveIBeenPwnedApiResponse = {
  headers: emptyHeaders,
  status: 404,
};

/**
 * This response has unique behavior. For some reason, the API includes an
 * object in the response body for this one, containing a human-readable
 * message. Manually populating the message here purely for use in tests.
 *
 * @internal
 */
export const TOO_MANY_REQUESTS: HaveIBeenPwnedApiResponse = {
  headers: new Map([['retry-after', '2']]),
  status: 429,
  body: {
    statusCode: 429,
    message: 'Rate limit is exceeded. Try again in 2 seconds.',
  },
};
