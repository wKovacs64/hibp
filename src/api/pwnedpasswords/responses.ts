/**
 * Known potential responses from the remote API.
 *
 * https://haveibeenpwned.com/api/v3#PwnedPasswords
 *
 */

/** @internal */
export interface PwnedPasswordsApiResponse {
  status: number;
  body?: string;
}

/** @internal */
export const OK: PwnedPasswordsApiResponse = {
  status: 200,
};

/** @internal */
export const BAD_REQUEST: PwnedPasswordsApiResponse = {
  status: 400,
  body: 'The hash prefix was not in a valid format',
};
