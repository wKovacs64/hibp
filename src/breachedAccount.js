import fetchFromApi from './internal/fetchFromApi';

/**
 * Fetches breach data for the specified account.
 *
 * @param {string} account a username or email address
 * @param {Object} [options] a configuration object
 * @param {string} [options.domain] a domain by which to filter the results
 * (default: all domains)
 * @param {boolean} [options.truncate] truncate the results to only include
 * the name of each breach (default: false)
 * @returns {Promise} a Promise which resolves to an array of breach objects
 * (or null if no breaches were found), or rejects with an Error
 * @example
 * hibp.breachedAccount('foo')
 *   .then((data) => {
 *     if (data) {
 *       // ...
 *     } else {
 *       // ...
 *     }
 *   })
 *   .catch((err) => {
 *     // ...
 *   });
 * @example
 * hibp.breachedAccount('bar', { truncate: true })
 *   .then((data) => {
 *     if (data) {
 *       // ...
 *     } else {
 *       // ...
 *     }
 *   })
 *   .catch((err) => {
 *     // ...
 *   });
 * @example
 * hibp.breachedAccount('baz', { domain: 'adobe.com', truncate: true })
 *   .then((data) => {
 *     if (data) {
 *       // ...
 *     } else {
 *       // ...
 *     }
 *   })
 *   .catch((err) => {
 *     // ...
 *   });
 * @memberof hibp
 * @function breachedAccount
 */
export default (account, options = {}) => {
  const endpoint = `/breachedaccount/${encodeURIComponent(account)}?`;
  const params = [];
  if (options.domain) {
    params.push(`domain=${encodeURIComponent(options.domain)}`);
  }
  if (options.truncate) {
    params.push('truncateResponse=true');
  }
  return fetchFromApi(`${endpoint}${params.join('&')}`);
};
