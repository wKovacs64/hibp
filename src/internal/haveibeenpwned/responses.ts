/**
 * Known potential responses from the remote API.
 *
 * Unfortunately, the API does not send a decent human-readable message back
 * with each response, but they are documented on the website:
 * https://haveibeenpwned.com/api/v2#ResponseCodes
 *
 * These objects simply provide a mapping between the HTTP response status code
 * and the corresponding human-readable message so we can throw a more
 * descriptive error for the consumer. (They are also leveraged in our tests.)
 */

/** @internal */
export const OK = {
  status: 200,
};

/** @internal */
export const BAD_REQUEST = {
  status: 400,
  statusText:
    'Bad request â€” the account does not comply with an acceptable format.',
};

/** @internal */
export const FORBIDDEN = {
  status: 403,
  statusText: 'Forbidden - no user agent has been specified in the request.',
};

/** @internal */
export const NOT_FOUND = {
  status: 404,
};

/**
 * This response has unique behavior. For some reason, the API includes a
 * human-readable message in the response body for this one. Manually populating
 * the message here purely for use in tests.
 *
 * @internal
 */
export const TOO_MANY_REQUESTS = {
  status: 429,
  data:
    'Rate limit exceeded, refer to acceptable use of the API: ' +
    'https://haveibeenpwned.com/API/v2#AcceptableUse',
};
