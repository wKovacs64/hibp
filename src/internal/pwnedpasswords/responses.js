/**
 * Known potential responses from the remote API.
 *
 * https://haveibeenpwned.com/API/v2#PwnedPasswords
 *
 */

export const RANGE_OK = {
  status: 200,
};

export const RANGE_BAD_REQUEST = {
  status: 400,
  response: 'The hash prefix was not in a valid format',
};
