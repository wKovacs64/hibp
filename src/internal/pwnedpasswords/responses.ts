/**
 * Known potential responses from the remote API.
 *
 * https://haveibeenpwned.com/API/v2#PwnedPasswords
 *
 */

export interface PwnedPasswordsApiResponse {
  // eslint-disable-next-line no-restricted-globals
  status: number;
  data?: string;
}

/** @internal */
export const OK = {
  status: 200,
};

/** @internal */
export const BAD_REQUEST = {
  status: 400,
  data: 'The hash prefix was not in a valid format',
};
