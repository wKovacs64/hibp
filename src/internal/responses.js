/**
 * Known potential responses from the remote API.
 *
 * Unfortunately, the API does not send a decent human-readable message back
 * with each response, but they are documented on the website:
 * https://haveibeenpwned.com/api/v2#ResponseCodes
 *
 * These objects simply provide a mapping between the HTTP response status code
 * and the corresponding human-readable message so we can throw a more
 * descriptive error for the consumer. (They are also leveraged in our tests.)
 */

export const OK = {
  status: 200,
};

export const BAD_REQUEST = {
  status: 400,
  statusText: 'Bad request â€” the account does not comply with an acceptable ' +
    'format.',
};

export const FORBIDDEN = {
  status: 403,
  statusText: 'Forbidden - no user agent has been specified in the request.',
};

export const NOT_FOUND = {
  status: 404,
  statusText: 'Not found - the account could not be found and has therefore ' +
    'not been pwned.',
};

/**
 * This response has unique behavior. For some reason, the API includes a
 * human-readable message in the response body for this one. Manually populating
 * the message here purely for use in tests.
 */
export const TOO_MANY_REQUESTS = {
  status: 429,
  response: 'Rate limit exceeded, refer to acceptable use of the API: ' +
    'https://haveibeenpwned.com/API/v2#AcceptableUse',
};
