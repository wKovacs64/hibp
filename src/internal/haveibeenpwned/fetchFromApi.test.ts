import AxiosError from 'AxiosError';
import breachedAccount from 'breachedAccount';
import dataClasses from 'dataClasses';
import { BAD_REQUEST, FORBIDDEN, TOO_MANY_REQUESTS } from './responses';
import mockAxios from './axiosInstance';

describe('internal (haveibeenpwned): fetchFromApi', () => {
  describe('request failure', () => {
    it('re-throws request setup errors', () => {
      const ERR = new Error('Set sail for fail!');
      mockAxios.get.mockRejectedValueOnce(ERR);
      expect(dataClasses()).rejects.toEqual(ERR);
    });
  });

  describe('invalid account format', () => {
    it('throws a "Bad Request" error', () => {
      mockAxios.get.mockRejectedValueOnce(new AxiosError(BAD_REQUEST));
      expect(breachedAccount('bad request')).rejects.toMatchSnapshot();
    });
  });

  describe('invalid request header', () => {
    it('throws a "Forbidden" error', () => {
      mockAxios.get.mockRejectedValueOnce(new AxiosError(FORBIDDEN));
      expect(breachedAccount('forbidden')).rejects.toMatchSnapshot();
    });
  });

  describe('rate limited', () => {
    it('throws a "Too Many Requests" error', () => {
      mockAxios.get.mockRejectedValueOnce(new AxiosError(TOO_MANY_REQUESTS));
      expect(breachedAccount('rate limited')).rejects.toMatchSnapshot();
    });
  });

  describe('unexpected HTTP error', () => {
    it('throws an error with the response status text', () => {
      mockAxios.get.mockRejectedValueOnce(
        new AxiosError({
          status: 999,
          statusText: 'Unknown - something unexpected happened.',
        }),
      );
      expect(breachedAccount('unknown response')).rejects.toMatchSnapshot();
    });
  });
});
