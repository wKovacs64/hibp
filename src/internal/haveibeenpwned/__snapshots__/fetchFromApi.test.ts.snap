// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`internal (haveibeenpwned): fetchFromApi invalid account format throws a "Bad Request" error 1`] = `[Error: Bad request â€” the account does not comply with an acceptable format.]`;

exports[`internal (haveibeenpwned): fetchFromApi invalid request header throws a "Forbidden" error 1`] = `[Error: Forbidden - no user agent has been specified in the request.]`;

exports[`internal (haveibeenpwned): fetchFromApi rate limited throws a "Too Many Requests" error 1`] = `[Error: Rate limit exceeded, refer to acceptable use of the API: https://haveibeenpwned.com/API/v2#AcceptableUse]`;

exports[`internal (haveibeenpwned): fetchFromApi unexpected HTTP error throws an error with the response status text 1`] = `[Error: Unknown - something unexpected happened.]`;
